core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java

```diff
@@ -267,31 +279,57 @@ public AtomicReference<Long> getWorkerId() {
   }
 
   @Override
-  public UfsStatus[] listStatus(String path, ListOptions options) throws IOException {
-    UfsStatus[] statuses = mListStatusCache.getIfPresent(path);
-    if (statuses == null) {
-      // Not found in cache. Query the Under File System.
-      statuses = mUfs.listStatus(path, options);
-
-      if (statuses == null) {
-        // If empty, the request path might be a regular file/object. Let's retry getStatus().
-        try {
-          UfsStatus status = mUfs.getStatus(path);
-          // listStatus() expects relative name to the @path.
-          status.setName("");
-          statuses = new UfsStatus[1];
-          statuses[0] = status;
-        } catch (FileNotFoundException e) {
-          statuses = null;
-        }
+  public UfsStatus[] listStatus(String path, ListStatusPOptions options) throws IOException {
+    final long syncIntervalMs = options.hasCommonOptions()
+        ? (options.getCommonOptions().hasSyncIntervalMs()
+        ? options.getCommonOptions().getSyncIntervalMs() : -1) :
+        -1;
+
+    final UfsStatus[] cachedStatuses;
+    final ListStatusResult resultFromCache = mListStatusCache.getIfPresent(path);
+    if (resultFromCache == null) {
+      cachedStatuses = null;
+    } else {
+      // Metadata is cached. Check if it is expired.
+      if (syncIntervalMs >= 0
+          && System.nanoTime() - resultFromCache.mTimeStamp > syncIntervalMs * Constants.MS_NANO) {
+        // The metadata is expired. Remove it from in-memory cache.
+        mListStatusCache.invalidate(path);
+        cachedStatuses = null;
+      } else {
+        // Cache is still valid. Use cached statuses.
+        cachedStatuses = resultFromCache.mUfsStatuses;
       }
+    }
+    if (cachedStatuses != null) {
+      return cachedStatuses;
+    }
 
-      // Add this into cache. Return value might be null if not found.
-      if (statuses != null) {
-        mListStatusCache.put(path, statuses);
+    // Not found in cache. Query the Under File System.
+    ListOptions ufsListOptions = ListOptions.defaults().setRecursive(
+        options.hasRecursive() ? options.getRecursive() : false);
+    UfsStatus[] freshStatusesFromUfs = mUfs.listStatus(path, ufsListOptions);
```

Comment:
Local variable `freshStatusesFromUfs` is reused.
"Reusing local variables" is just a suggestion, not mandatory.

Clearly, here reusing the local variable `freshStatusesFromUfs ` will make it easier to understand the code.

Reply:
Yes, I second that. Both `ufs.listStatus()` and `ufs.getStatus()` calls give you fresh UFS statuses, they are just 2 different ways to do the same thing. Putting either result into `freshStatusesFromUfs` does not go against what its name suggest.

Reply:
Thank you, @dbw9580